lib/IDE/SourceEntityWalker.cpp

```diff
@@ -762,6 +771,54 @@ bool SemaAnnotator::handleCustomAttributes(Decl *D) {
   return true;
 }
 
+bool SemaAnnotator::handleClosureAttributes(ClosureExpr *E) {
+  ModuleDecl *MD = E->getParentModule();
+  for (auto *customAttr : E->getAttrs().getAttributes<CustomAttr, true>()) {
+    SourceFile *SF =
+        MD->getSourceFileContainingLocation(customAttr->getLocation());
+    ASTNode expansion = SF ? SF->getMacroExpansion() : nullptr;
+    if (!shouldWalkMacroArgumentsAndExpansion().second && expansion) {
+      continue;
+    }
+
+    if (auto *Repr = customAttr->getTypeRepr()) {
+      if (!Repr->walk(*this)) {
+        return false;
+      }
+    }
+
+    if (auto *Args = customAttr->getArgs()) {
+      if (!Args->walk(*this)) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+bool SemaAnnotator::handleTypeAttributes(AttributedTypeRepr *T) {
```

Comment:
@bnbarham In the other two functions above, we're able to get to the `ModuleDecl` to check whether we're inside a macro expansion, but I wasn't able to figure out how to do that here from a `TypeRepr` (or from its associated `TypeExpr`). So I'm not sure if this is completely correct.

Reply:
I would expect that we stop walking a macro expansion if `shouldWalkMacroArgumentsAndExpansion` on one of the parent nodes already, so we don’t need to check here. Would like to hear @bnbarham’s thoughts on it though, I might be missing something.

Reply:
I would expect that to be true in both cases. We're checking the general custom attribute case as a macro could add an attribute and thus that attribute could be within a macro. But for these two I can't think of a case where that could happen. CC @hamishknight in case he can think of a case that we're missing 😅

Reply:
Yeah I don't think we need to check in either case, macros can only add attributes to decls, so for a closure or TypeRepr attribute to be in a macro expansion there would need to be some parent in the expansion too, which we'd apply the walking behavior to.